src/Sylius/Bundle/PromotionBundle/DependencyInjection/SyliusPromotionExtension.php

```diff
@@ -55,14 +72,18 @@ public function load(array $config, ContainerBuilder $container)
             ->addArgument(new Reference('sylius.repository.promotion'))
         ;
         $container->setDefinition('sylius.factory.promotion_coupon', $decoratedCouponFactoryDefinition);
+    }
 
-        $container
-            ->getDefinition('sylius.form.type.promotion_action')
-            ->replaceArgument(1, new Reference('sylius.registry.promotion_action'))
-        ;
-        $container
-            ->getDefinition('sylius.form.type.promotion_rule')
-            ->replaceArgument(1, new Reference('sylius.registry.promotion_rule_checker'))
-        ;
+
+    /**
+     * @param ContainerBuilder $container
+     */
+    private function overwriteActionFactory(ContainerBuilder $container)
+    {
+        $oldActionFactory = $container->getDefinition('sylius.factory.promotion_action');
+        $newActionFactoryDefinition = new Definition(ActionFactory::class);
+
+        $actionFactory = $container->setDefinition('sylius.factory.promotion_action', $newActionFactoryDefinition);
+        $actionFactory->addArgument($oldActionFactory);
```

Comment:
The `ActionFactory::class` should be configured in the `Configuration` class and the overwrite should look like this:
```
$baseFactoryDefinition = new Definition(Factory::class, [new Parameter('sylius.model.promotion_action.class')]);
$promotionActionFactoryClass = $container->getParameter('sylius.factory.promotion_action.class');
$decoratedPromotionActionFactoryDefinition = new Definition($promotionActionFactoryClass, [$baseFactoryDefinition]);

$container->setDefinition('sylius.factory.promotion_action', $decoratedPromotionActionFactoryDefinition);
```